meshmode/dof_array.py

```diff
@@ -114,96 +118,180 @@ def from_list(cls, actx: Optional[ArrayContext], res_list) -> "DOFArray":
         :arg actx: If *None*, the arrays in *res_list* must be
             :meth:`~meshmode.array_context.ArrayContext.thaw`\ ed.
         """
+        from warnings import warn
+        warn("DOFArray.from_list is deprecated and will disappear in 2021.",
+                DeprecationWarning, stacklevel=2)
         if not (actx is None or isinstance(actx, ArrayContext)):
             raise TypeError("actx must be of type ArrayContext")
 
-        ngroups = len(res_list)
+        return cls(actx, tuple(res_list))
 
-        result = np.empty(ngroups, dtype=object).view(cls)
-        result.array_context = actx
+    # }}}
 
-        # 'result[:] = res_list' may look tempting, however:
-        # https://github.com/numpy/numpy/issues/16564
-        for igrp in range(ngroups):
-            result[igrp] = res_list[igrp]
+    # {{{ sequence protocol
 
-        return result
+    def __len__(self):
+        return len(self._data)
 
-    # {{{ work around numpy failing to compare obj arrays of arrays
+    def __getitem__(self, i):
+        return self._data[i]
 
-    def _comparison(self, operator_func, other):
-        from numbers import Number
-        if isinstance(other, DOFArray):
-            return obj_array_vectorize_n_args(operator_func, self, other)
+    def __iter__(self):
+        return iter(self._data)
 
-        elif isinstance(other, Number):
-            return obj_array_vectorize(
-                    lambda self_entry: operator_func(self_entry, other),
-                    self)
+    # }}}
 
+    @property
+    def shape(self):
+        return (len(self),)
+
+    def _like_me(self, data):
+        return DOFArray(self.array_context, tuple(data))
+
+    def _bop(self, f, op1, op2):
+        """Broadcasting logic for a generic binary operator."""
+        if isinstance(op1, DOFArray) and isinstance(op2, DOFArray):
+            if len(op1._data) != len(op2._data):
+                raise ValueError("DOFArray objects in binary operator must have "
+                        f"same length, got {len(op1._data)} and {len(op2._data)}")
+            return self._like_me([
+                f(op1_i, op2_i)
+                for op1_i, op2_i in zip(op1._data, op2._data)])
```

Comment:
```suggestion
            return self._like_me(f(op1_i, op2_i) for op1_i, op2_i in zip(op1._data, op2._data))
```
Does this work? It gets cast to a tuple anyway.

Reply:
My thought was that list comprehensions are probably minimally faster than generator expression. But I have exactly zero data to back that up.

Reply:
I was thinking the other way around! Something to `timeit` for peace of mind, I guess.

Reply:
You're right! At least for [this contrived example](https://gist.github.com/alexfikl/1c9fef4bade9a22ed445b25cfeffadc3).

TIL

Reply:
Thanks for checking!